### ISSUE TITLE
Make datasource provider dynamic via environment variables

### ISSUE THREAD

**Issue by schickling** [prisma/prisma#1487]
## Context

It's a common use case to use different but very similar databases for different environments (e.g. using SQLite for local development and using Postgres in production). Prisma is especially well suited to support this use case as every Prisma connector is aware of the database's capabilities and can therefore safely (i.e. at schema validation / client generation time) determine whether two databases are compatible. It does so by reducing the supported database capabilities down to the "overlapping capabilities" (or in mathematical terms: "Lowest common denominator"). For example even though Postgres supporting scalar list fields, when a user wants to use Postgres and SQLite in the same project, the scalar list feature won't be available i.e. the schema won't validate and client operations won't be generated.

In the past we tried to solve this by supporting an optional `enabled: env("SOME_VAR")` (see e.g. #1142) however this solution had multiple problems such as: 

- It's unintuitive how `enabled` works and how it's supposed to be used. I.e. `$SOME_VAR` is casted to a boolean
- It's also not clear in which cases the user should configure multiple datasources for multiple environments or just replace environment variables

## Proposal

Here is a quick summary of what I'm proposing. More details afterwards.

1. Remove the `provider: "postgresql"` property from the `datasource` block as it's redundant to the information contained in the `url` property.
2. This enables datasources to become fully dynamic as they can be set via environment variables (and even allows for different datasource types e.g. dynamically switching between SQLite and Postgres).
3. To still enable static schema validation, we'll introduce a `providers: ["postgresql", "sqlite"]` property

### Details

⚠️ Still WIP

## Related issues

- This allows affects (and simplifies) #1480
- Further simplifies #1142
- Enables https://github.com/prisma/prisma-client-js/issues/184

**Comment by Errorname** [comment]
Hi there, just dropping by to say that I'm really looking for this feature to be implemented for my use case (SQLite in dev and Postgres in prod) 🙏

**Comment by ryands17** [comment]
A use case by a user referenced from [this] discussion

The use case is the following:
One separate db/schema per project (could be whatever) with exactly the same set of tables.
One common db for common data.
There are not hundreds of projects or that would become really cumbersome (even though things can be automated).

So there is a mix of multi-tenant need (same set of tables over multiple db) and multiple database with different tables.

> Just for information - this is usually easily solved by existing orm:

TypeOrm has direct support for multiple connections (getConnection("cx1").getRepository...
mikro-orm does not have specific support for multiple connections but you can instantiate multiple.
For both of them, you link entities to each connection so that you don't have to repeat entity definition for the situation when you have the same set of tables.
I looked at the issue you are mentioning but it doesn't look like the same matter. As far as I could understand the request in this issue is for being able to switch the type of database depending on the environment (for instance being able to use Sqlite in dev while using posgres in prod).

Looking at your schema, I believe that could be easily achieved (well kind of) on the schema side of things:

datasource proj01 {
  provider = "postgresql"
  url      = env("DATABASE_URL_PROJ01")
  models = [Post, User]
}

datasource proj02 {
  provider = "postgresql"
  url      = env("DATABASE_URL_PROJ02")
  models = [Post, User]
}

datasource common {
  provider = "postgresql"
  url      = env("DATABASE_URL_COMMON")
  models = [Config]
}



generator client {
  provider = "prisma-client-js"
}
model Post {
  id        Int     @id @default(autoincrement())
  title     String
  content   String?
  published Boolean @default(false)
  author    User?   @relation(fields:  [authorId], references: [id])
  authorId  Int?
}
model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
  posts Post[]
}

model Config {
  id    Int     @id @default(autoincrement())
  field string
}
You would then need to make it possible to use one connection/db or another:

import { PrismaClient } from '@prisma/client'
const proj01= new PrismaClient('proj01')
const proj02= new PrismaClient('proj02')
const proj01Users = await proj01.user.findMany()
const proj02Users = await proj02.user.findMany()
Even for those apps that are based around one single schema, it happens here and there that there is the need to access some other database.

**Comment by thedavidprice** [comment]
@janpio Great to see this is labeled for the current/next Beta release! 😉

It might be worth a brief conversation with both @cannikin and me to connect some dots between this "dynamic provider" feature, a recent Redwood [deployment Issue](https://github.com/redwoodjs/redwood/issues/469) (resolved with a plugin workaround), and the corresponding Prisma Migrate Issue https://github.com/prisma/migrate/issues/433 

Thanks again!

**Comment by FredericLatour** [comment]
I opened a separate issue for handling multiple connections/database: 
* #2443 

What I'm proposing in this issue would solve the matter exposed in this issue as well.

**Comment by shainegordon** [comment]
why did this work in for example `preview22`, but not in the current beta?

**Comment by pantharshit00** [comment]
@shainegordon Most likely because we added a ton of validation logic for the schema now.

**Comment by BjoernRave** [comment]
Just out of interest, can you say when this will land? We are waiting on this for `prisma-multi-tenant` :)

**Comment by FredericLatour** [comment]
@BjoernRave If I understood correctly the information [provided here](https://github.com/prisma/prisma/issues/2443#issuecomment-630679118), it's already possible to do multi-tenant.
I'm getting confused now ;)

**Comment by BjoernRave** [comment]
@FredericLatour yes, it is possible. The problem is being able to support multiple datasources in a library. @Errorname also managed to make it work, but it's not very nice.

**Comment by FredericLatour** [comment]
@BjoernRave Do you mean multiple data sources with different models in each datasource?

Because, unless I misunderstood something, in a pure multi-tenant scenario, the following should work rather nicely:

```typescript
import { PrismaClient } from '@prisma/client'

const client1 = new PrismaClient({ datasources: { mydb: 'postgres://localhost/db1' } })
const client2 = new PrismaClient({ datasources: { mydb: 'postgres://localhost/db2' } })
```

**Comment by Errorname** [comment]
@FredericLatour the tool [prisma-multi-tenant](https://github.com/Errorname/prisma-multi-tenant) adds some features over Prisma:
- Access multiple databases (using kind of the same trick as you are suggesting)
- Create "Management" database that keeps tracks of all of your tenants (using a specific schema)
- Keep your tenants databases migrations steps in sync (using prisma migrate)
- Easily create and delete tenants either by CLI or programatically (using a combination of previous tricks)

There is another feature that I absolutely want for prisma-multi-tenant:
- Use an sqlite db for dev, then use a postgresql db for prod, without changing anything;
- Or, use a mysql db for management, and postgresql db for tenants;
- Or, some tenants on postgresql, and some other on mysql; etc..

At the moment, I am using some tricks to make this work, but it's really instable, and breaks on nearly every release. Having the "dynamic datasource provider" as a feature of Prisma would definitely go a long way into make prisma-multi-tenant a stable library for production usage

**Comment by FredericLatour** [comment]
@Errorname your project looks great. I don't have the need for the features you are providing and, as far as I am concerned, the less moving parts the better.

Even though, I prefer the solution I proposed because I believe it can solve any scenario, why would you not be able to use this approach as proposed by Prisma (I didn't try personnaly):

```typescript
import { PrismaClient } from '@prisma/client'

const client1 = new PrismaClient({ datasources: { mydb: 'postgres://localhost/db1' } })
const client2 = new PrismaClient({ datasources: { mydb: 'sqlite://...' } })
```

**Comment by Errorname** [comment]
@FredericLatour Currently, when generating a Prisma Client, it is specific to a provider (sqlite, postgres or mysql). You cannot reuse the same client and specify a provider which it wasn't generated for. But there is a feature request (this very issue 😄) to do that.

As a workaround, I need to have 3 different datasources for each provider, and enable/disable them based on what provider the user wants for its management datasources.

**Comment by FredericLatour** [comment]
@Errorname But I guess it would be difficult to have a dynamic provider if they use provider specific features at generation time. Or am I missing something?

**Comment by Errorname** [comment]
@FredericLatour As this issue states:
> Prisma is **especially well suited to support this use case** as every Prisma connector is aware of the database's capabilities and can therefore safely (i.e. at schema validation / client generation time) determine whether two databases are compatible. It does so by reducing the supported database capabilities down to the "overlapping capabilities" (or in mathematical terms: "Lowest common denominator"). For example even though Postgres supporting scalar list fields, when a user wants to use Postgres and SQLite in the same project, the scalar list feature won't be available i.e. the schema won't validate and client operations won't be generated.

I'm sure it is not an easy feature (I'm only a user of Prisma, I don't know too much about database stuff, so I can't judge), but they seem to be confident that they are able to make it happen

**Comment by FredericLatour** [comment]
@Errorname I'm not sure I like their approach of having a downgraded feature set, just because you would like using sqlite in dev or for tests. That said it may have it's use but I stand on my solution here. It's way easier and solve other concerns as well:

For instance, using my approach you would have additional datasources but by using `models ` property, you don't have to repeat model definitions. You wouldn't have to generate multiple schemas as well.


```typescript
datasource tenant {
  provider = "postgresql"
  url      = env("DATABASE_URL_TO_SOME_TENANT")
  models = [Post, User]
}

datasource tenant-dev {
  provider = "sqlite"
  url      = env("DATABASE_URL_TO_SOME_SQLITE_TENANT")
  models = [Post, User]
}

datasource common {
  provider = "postgresql"
  url      = env("DATABASE_URL_COMMON")
  models = [Config]
}

datasource common-dev {
  provider = "sqlite"
  url      = env("DATABASE_URL_COMMON")
  models = [Config]
}

generator client {
  provider = "prisma-client-js"
}
model Post {
  id        Int     @id @default(autoincrement())
  title     String
  content   String?
  published Boolean @default(false)
  author    User?   @relation(fields:  [authorId], references: [id])
  authorId  Int?
}
model User {
  id    Int     @id @default(autoincrement())
  email String  @unique
  name  String?
  posts Post[]
}

model Config {
  id    Int     @id @default(autoincrement())
  field string
}
```

You would then need to make it possible to use one connection/db or another:

```typescript
import { PrismaClient } from '@prisma/client'
const [tenant, common]= process.env.NODE_ENV === prod ? ['tenant', 'common'] : ['tenant-dev', common-dev]

const someTenant = new PrismaClient(tenant, { datasources: { mydb: <tenant-x-url } })
const commonDb = new PrismaClient(common)

etc ...
```

**Comment by mavilein** [comment]
We would like to move forward with this proposal now. I revisited it and would like to suggest a slightly different proposal for this change.

## Summary
1. Extend the Prisma schema syntax to allow the `provider` property in `datasource` blocks to also accept an array of values, e.g. `provider: ["postgresql", "sqlite"]`.
2. The active provider for the datasource is then determined by the provided `url` property.
3. The schema validation must determine the set of capabilities supported by all specified providers. Only capabilities that are supported by all providers can be then used. This requires the `provider` property to be always specified as opposed to the initial proposal.

## Details


### 1. syntax changes
The current syntax for `provider` is still supported as is:
```prisma
datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}
```

Additionally we will allow to also provide an array of providers. This is contrary to the original proposal that suggested to add the plural form `providers`. By only extending the singular `provider` property makes the change simpler by not dealing with edge cases such as both properties being specified.
```prisma
datasource db {
  provider = ["sqlite", "postgres"]
  url      = env("DATABASE_URL")
}
```

### 2. Process for determining the active provider
If multiple providers are specified the active provider needs to be determined in order to be able to interact with the underlying database. This is done by the following process:

1. Resolve the value for the `url` property of the datasource. Either it is a static value or an environment variable that needs to be resolved.
2. Loop over the providers specified in the `provider` property. The resolved URL is passed to each provider in a function called `canHandleURL`. The provider must return true or false depending on whether it can handle this URL. E.g. the `postgres` provider would return true for all URLs starting with `postgres://`. 
3. The first provider that returns true becomes the active provider.

This logic implies that connection strings for different providers must **not** be overlapping. Consider for instance `provider: ["cockroach", "postgres"]` that uses Postgres and CockroachDB. Cockroach DB is wire compatible with Postgres and hence can deal with Postgres connection URLs like `postgres://`. Therefore the described logic would always determine the first provider in the array as active. That means that we need to be able to differentiate our provider URLs in such cases by e.g. adding prefixes like `cockroach+postgres://`.


### 3. Influence of capabilities on schema validation

If multiple providers are specified the Prisma schema needs to be compatible with all of them. For example when using `provider: ["sqlite", "postgres"]` the schema needs to be compatible with Postgres and SQLite. The consequence in that case is that scalar list fields, e.g. `tags String[]`, would not be allowed as Postgres supports them but SQLite does not. This example illustrates that **we need to determine the set of capabilities supported by all of the specified providers**. A capability is only supported by a set of multiple providers if each supports that capability.
Following is an overview of the capabilities that the current set of available providers exposes. For each capability it is noted whether a given connector supports it or not. The list of capabilities is going to grow when more providers or features (like custom types) are added.

| capabilities                     | sqlite | postgres | mysql |
|----------------------------------|--------|----------|-------|
| scalar lists                     | no     | yes      | no    |
| enums                            | no     | yes      | no    |
| json                             | no     | yes      | yes   |
| relations over non unique fields | no     | no       | yes   |

**Comment by wanseob** [comment]
> Loop over the providers specified in the provider property. The resolved URL is passed to each provider in a function called canHandleURL. The provider must return true or false depending on whether it can handle this URL. E.g. the postgres provider would return true for all URLs starting with postgres://.

This looks simple and easy to use!

**Comment by timsuchanek** [comment]
This is implemented in the latest dev!
You can try it out with `@prisma/cli@dev`.

**Comment by softmarshmallow** [comment]
> We would like to move forward with this proposal now. I revisited it and would like to suggest a slightly different proposal for this change.
> 
> ## Summary
> 1. Extend the Prisma schema syntax to allow the `provider` property in `datasource` blocks to also accept an array of values, e.g. `provider: ["postgresql", "sqlite"]`.
> 2. The active provider for the datasource is then determined by the provided `url` property.
> 3. The schema validation must determine the set of capabilities supported by all specified providers. Only capabilities that are supported by all providers can be then used. This requires the `provider` property to be always specified as opposed to the initial proposal.
> 
> ## Details
> ### 1. syntax changes
> The current syntax for `provider` is still supported as is:
> 
> ```prisma
> datasource db {
>   provider = "sqlite"
>   url      = env("DATABASE_URL")
> }
> ```
> 
> Additionally we will allow to also provide an array of providers. This is contrary to the original proposal that suggested to add the plural form `providers`. By only extending the singular `provider` property makes the change simpler by not dealing with edge cases such as both properties being specified.
> 
> ```prisma
> datasource db {
>   provider = ["sqlite", "postgres"]
>   url      = env("DATABASE_URL")
> }
> ```
> 
> ### 2. Process for determining the active provider
> If multiple providers are specified the active provider needs to be determined in order to be able to interact with the underlying database. This is done by the following process:
> 
> 1. Resolve the value for the `url` property of the datasource. Either it is a static value or an environment variable that needs to be resolved.
> 2. Loop over the providers specified in the `provider` property. The resolved URL is passed to each provider in a function called `canHandleURL`. The provider must return true or false depending on whether it can handle this URL. E.g. the `postgres` provider would return true for all URLs starting with `postgres://`.
> 3. The first provider that returns true becomes the active provider.
> 
> This logic implies that connection strings for different providers must **not** be overlapping. Consider for instance `provider: ["cockroach", "postgres"]` that uses Postgres and CockroachDB. Cockroach DB is wire compatible with Postgres and hence can deal with Postgres connection URLs like `postgres://`. Therefore the described logic would always determine the first provider in the array as active. That means that we need to be able to differentiate our provider URLs in such cases by e.g. adding prefixes like `cockroach+postgres://`.
> 
> ### 3. Influence of capabilities on schema validation
> If multiple providers are specified the Prisma schema needs to be compatible with all of them. For example when using `provider: ["sqlite", "postgres"]` the schema needs to be compatible with Postgres and SQLite. The consequence in that case is that scalar list fields, e.g. `tags String[]`, would not be allowed as Postgres supports them but SQLite does not. This example illustrates that **we need to determine the set of capabilities supported by all of the specified providers**. A capability is only supported by a set of multiple providers if each supports that capability.
> Following is an overview of the capabilities that the current set of available providers exposes. For each capability it is noted whether a given connector supports it or not. The list of capabilities is going to grow when more providers or features (like custom types) are added.
> 
> capabilities	sqlite	postgres	mysql
> scalar lists	no	yes	no
> enums	no	yes	yes
> json	no	yes	yes
> relations over non unique fields	no	no	yes
> ### Edge cases
> We have to be aware that `prisma introspect` might yield invalid schemas with this proposal. Consider a schema using `provider: ["sqlite", "postgres"]`. Depending on the env var the postgres datasource might become active. The underlying database might use a JSON column. The introspection would then add the JSON field. Subsequent schema validations would fail because JSON is not part of the set of supported capabilities for those providers.
> But this can only happen when starting with an existing database. So that is probably a good trade off.

I think the most case we need multiple provider is for dev/prod environments. Meanwhile diminishing the capability of each database seems to be not matching the original reasons/needs for me.

Will there not be a way to specify dev provider so for the production (in my case postgres) postgres will have it's native compatibilities?

**Comment by fivethreeo** [comment]
For a quick fix

```bash
sed -ie 's/postgresql/sqlite/g' prisma/schema.prisma 
sed -ie 's/sqlite/postgresql/g' prisma/schema.prisma 
```

**Comment by yanickrochon** [comment]
I do not understand the resolution for this; how do I set a SQLite provider for dev environment and PostgreSQL for production? It seems like it was implemented, then removed?

**Comment by cannikin** [comment]
In the "original" builds of Prisma v2 you could swap out your DB provider based on environment and use SQLite in dev and then Postgres or MySQL in production. But they re-wrote their migration engine and made it incompatible with doing that. The migrations themselves are just SQL code, and they are specific to the DB provider, so now you're pretty much stuck using a single provider in all environments. :(

**Comment by yanickrochon** [comment]
Well, that is lame.

**Comment by JakubKoralewski** [comment]
> The migrations themselves are just SQL code, and they are specific to the DB provider, so now you're pretty much stuck using a single provider in all environments. :(

But I could just `prisma db push` on the development SQLite database :(

**Comment by bhuynhdev** [comment]
So how may I use this `provider` array feature? I am using prisma 4.5.0, and it is saying "Error validating datasource `db`: The provider argument in a datasource must be a string literal"

**Comment by shellscape** [comment]
I realize that this is a very old PR, but I wanted to mention that environment variables are a terrible way to store secrets, and that this implementation doesn't allow for using secret managers e.g. AWS SecretsManager.

**Comment by janpio** [comment]
We implemented what people are asking for and using in other tools.
For AWS Secrets Manager support, see (and upvote) https://github.com/prisma/prisma/issues/7534

**Comment by jakerich1** [comment]
For anyone who is implementing a dynamically changing provider property based on something like the node env in their own project. Be aware that the sqlite and postgresql provider will generate different types after running `npx prisma generate` Caught myself out today.

The solution I'll using will be to make use of the branching feature that Planetscale offer, so I can target a dev db with the exact same schema and provider.

I'd love a way of being able to use sqlite though, makes things a bit cheaper

**Comment by janpio** [comment]
Suggestions on how that could work with Migrations and also the types and general Client API?

**Comment by souzacavalheiro** [comment]
Hello everybody! I didn't see in recent comments any use case related to 'db for user group'.

exemplifying:

  userA => group1 => mysqldb1
  userb => group2 => mysqldb2

Our users are company employees and each one accesses their own database... same tables, same schema, same provider but different databases.
Any tips?

ps: translated by google

**Comment by janpio** [comment]
Are all these databases of the same type, so all MySQL or PostgreSQL? Then this issue here is not relevant. You do not need to use a dynamic database `provider` field - just change the connection string depending on the user. You probably want to have multiple installations of your app, or see how to implement "multi tenancy" with Prisma (use Google or search this repository).

**Comment by haydenk** [comment]
@souzacavalheiro 
> Hello everybody! I didn't see in recent comments any use case related to 'db for user group'.
> 
> exemplifying:
> 
> userA => group1 => mysqldb1 userb => group2 => mysqldb2
> 
> Our users are company employees and each one accesses their own database... same tables, same schema, same provider but different databases. Any tips?
> 
> ps: translated by google

https://www.prisma.io/docs/guides/other/multi-schema

It's still in preview so I wouldn't put too much trust in it yet but this is coming.

@janpio's solution is not the most ideal or elegant because then it means you have to create a whole new connection for a different schema but it's the only real solution available at the moment until multi-schema is out of preview.

**Comment by janpio** [comment]
@haydenk The multi schema preview feature is for distributing one Prisma schema across multiple database schemas, it does not help with the same database schema being on multiple databases or schemas - that it what I referred to with "multi tenancy".

**Comment by haydenk** [comment]
@janpio good point. I overlooked that it was different database servers.

**Comment by P-Info** [comment]
Thank you everyone for the responses. An example I found in another library 'lucidORM by:AdonisJs', for this situation...

https://docs.adonisjs.com/guides/database/introduction#switching-between-multiple-connections

The name of the bank that should be used in the query is informed in the query:

import Database from '@ioc:Adonis/Lucid/Database'

for 'single_databases' => Database.from('posts').select('*')

for 'multi_databases' => Database.connection('mysql_database_1').from('posts').select('*')

**Comment by souzacavalheiro** [comment]
Thank you everyone for the responses. An example I found in another library 'lucidORM by:Adonisjs', for this situation...

https://docs.adonisjs.com/guides/database/introduction#switching-between-multiple-connections

The name of the bank that should be used in the query is informed in the query:

import Database from '@ioc:Adonis/Lucid/Database'

for 'single_databases' => Database.from('posts').select('*')

for 'multi_databases' => Database.connection('mysql_database_server1').from('posts').select('*')

**Comment by janpio** [comment]
Please open your own discussion for this instead of commenting in an old, closed issue @souzacavalheiro. Thanks.

**Comment by alejandrobalderas** [comment]
> Hi there, just dropping by to say that I'm really looking for this feature to be implemented for my use cases (SQLite in dev and Postgres in prod & multi-tenancy accross different databases) 🙏

As far as I understood this was the whole essence of this feature request and exactly what I would try to achieve. It appears this was possible but it does not work anymore as far as I can tell. My question is what do we do now?

**Comment by 5knnbdwm** [comment]
@alejandrobalderas idk, seems more and more likely that I'm gonna move to other solutions over prisma.

**Comment by AndresRojoInteracso** [comment]
I found in google a easy to review implementation/example on how to have two different providers base on the enviroment, but it forces to have two schema.prisma files what i consider could lead to problems but maibe it is usefull to someone.
https://codesandbox.io/p/sandbox/prisma-multiple-database-providers-g4s3e

**Comment by wildfunctions** [comment]
Not using prisma over this.

**Comment by IArnaut2** [comment]
Prisma not being able to handle multiple databases is so stupid.
